Feat: Route Transformer Advanced plugin (#979)
Kong/developer.konghq.com: app/_includes/plugins/request-response-transformer/templates.md

columns:
  - title: Type
    key: type
  - title: Template
    key: template
rows:
  - type: Header
    template: |
      * `$(headers.<header-name>)`
      * `$(headers["<header-name>"])`
  - type: Query parameter
    template: |
      * `$(query_params.<query-param-name>)`
      * `$(query_params["<query-param-name>"])`
  - type: Captured URIs
    template: |
      * `$(uri_captures.<group-name>)`
      * `$(uri_captures["<group-name>"])`
  - type: Shared variables
    template: |
      * `$(shared.<variable-name>)`
      * `$(shared["<variable-name>"])`
    
{% endtable %}

To escape a template, wrap it inside quotes and pass inside another template.
For example:

```
$('$(something_that_needs_to_escaped)')
```

{:.info}
> **Note:** The plugin creates a non-mutable table of request headers, query strings, and captured URIs before transformation. Therefore, any update or removal of parameters used in a template does not affect the rendered value of a template.

### Advanced templates

The content of the placeholder `$(...)` is evaluated as a Lua expression, so
logical operators may be used. For example:

```
$(uri_captures["user-id"] or query_params["user"] or "unknown")
```

> This will first look for the path parameter named `user-id` (in `uri_captures`). If not found, it will
> return the query parameter named `user`. If that also doesn't exist, it returns the default
> value '"unknown"'.

Reviewer: ```suggestion
This example will look for parameters in the following order:
1. Looks for a path parameter named `user-id` in `uri_captures`.
2. If not found, it will return the query parameter named `user`.
3. If that also doesn't exist, it returns the default value `"unknown"`.
```

> Constant parts can be specified as part of the template outside the dynamic
> placeholders. For example, creating a basic-auth header from a query parameter
> called `auth` that only contains the base64-encoded part:

Reviewer: ```suggestion
Constant parts can be specified as part of the template outside the dynamic
placeholders. For example, you can create a basic-auth header from a query parameter
called `auth` that only contains the base64-encoded part:
```

```
Basic $(query_params["auth"])
```

Lambdas are also supported if wrapped as an expression like this:

```
$((function() ... implementation here ... end)())
```

> A complete Lambda example for prefixing a header value with `Basic` if it's not

Reviewer: ```suggestion
Here's a complete Lambda example for prefixing a header value with `Basic` if it's not
```

already included:

```
$((function()
    local value = headers.Authorization
    if not value then
      return
    end
    if value:sub(1, 6) == "Basic " then
      return value            -- was already properly formed
    end
    return "Basic " .. value  -- added proper prefix
  end)())
```

> {:.info}
> > **Note:** Especially in multi-line templates like the example above, make sure not
> to add any trailing white space or new lines. Because these would be outside the
> placeholders, they would be considered part of the template, and hence would be
> appended to the generated value.
> The environment is sandboxed, meaning that Lambdas will not have access to any
> library functions, except for the string methods (like `sub()` in the example
> above).

Reviewer: ```suggestion
The environment is sandboxed, meaning that Lambdas won't have access to any
library functions, except for the string methods (like `sub()` in this example).

{:.info}
> **Note:** Make sure not to add any trailing whitespace or newlines, especially in multi-line templates. 
These would be outside the placeholders and would be considered part of the template, and hence would be appended to the generated value.
```
The second half looks like it doesn't relate to the first part at all, and instead describes the example - so i moved it.
